src/NServiceBus.Core/Causation/AttachCausationHeadersBehavior.cs

```diff
@@ -44,11 +49,36 @@ static void SetConversationIdHeader(IOutgoingLogicalMessageContext context, Inco
 
             if (hasUserDefinedConversationId)
             {
-                // do not override user defined conversation id if no incoming message exists.
                 return;
             }
 
-            context.Headers[Headers.ConversationId] = CombGuid.Generate().ToString();
+            string conversationId;
+            bool userProvidedId;
+
+            try
+            {
+                userProvidedId = tryGetConversationIdDelegate(new ConversationIdStrategyContext(context.Message), out conversationId);
+            }
+            catch (Exception exception)
+            {
+                throw new Exception($"Failed to execute CustomConversationId delegate. This configuration option was defined using {nameof(EndpointConfiguration)}.{nameof(MessageCausationConfigurationExtensions.CustomConversationIdStrategy)}.", exception);
+            }
+
+            if (userProvidedId)
+            {
+                if (string.IsNullOrEmpty(conversationId))
```

Comment:
This feels messy to me. Why introduce a delegate type just to have an out parameter, so that you now have to check the bool and the value of the out parameter?